otter/log/cloudfeeds.py

```diff
@@ -0,0 +1,173 @@
+    """
+    def __init__(self, message):
+        self.unsuitable_message = message
+
+
+# Mapping of items in a log event to cloud feeds event
+log_cf_mapping = {
+    "scaling_group_id": "scalingGroupId",
+    "policy_id": "policyId",
+    "webhook_id": "webhookId",
+    "username": "username",
+    "desired_capacity": "desiredCapacity",
+    "current_capacity": "currentCapacity",
+    "message": "message"
+}
+
+
+def sanitize_event(event):
+    """
+    Sanitize event by removing all items except the ones in autoscale schema.
+
+    :param dict event: Event to sanitize as given by Twisted
+
+    :return: (dict, bool, str) tuple where dict -> sanitized event,
+        bool -> is it error event?, str -> ISO8601 formatted UTC time of event
+    """
+    cf_event = {}
+    error = False
+
+    # format message
+    event_copy = deepcopy(event)
+    PEP3101FormattingWrapper(lambda e: None)(event_copy)
+    msg = event_copy["message"]
+    try:
+        event["message"] = msg[0]
+    except KeyError:
+        raise UnsuitableMessage("No message in event")
+
+    # map keys in event to CF keys
+    for log_key, cf_key in log_cf_mapping.iteritems():
+        if log_key in event:
+            cf_event[cf_key] = event[log_key]
+
+    if event.get('isError', False):
+        error = True
+        if ('traceback' in cf_event['message'] or
+           'exception' in cf_event['message']):
+            raise UnsuitableMessage(cf_event['message'])
+
+    return (cf_event, error,
+            datetime.utcfromtimestamp(event["time"]).isoformat())
+
+
+request_format = {
+    "entry": {
+        "@type": "http://www.w3.org/2005/Atom",
+        "title": "autoscale",
+        "content": {
+            "event": {
+                "@type": "http://docs.rackspace.com/core/event",
+                "id": "",
+                "version": "2",
+                "eventTime": "",
+                "type": "INFO",
+                "region": "",
+                "product": {
+                    "@type": "http://docs.rackspace.com/event/autoscale",
+                    "serviceCode": "Autoscale",
+                    "version": "1",
+                    "message": ""
+                }
+            }
+        }
+    }
+}
+
+
+def prepare_request(req_fmt, event, error, timestamp, region,
+                    _uuid=uuid.uuid4):
+    """
+    Prepare request based on request format
+    """
+    request = deepcopy(req_fmt)
+    if error:
+        request['entry']['content']['event']['type'] = 'ERROR'
+    request['entry']['content']['event']['region'] = region
+    request['entry']['content']['event']['eventTime'] = timestamp
+    request['entry']['content']['event']['product'].update(event)
+    request['entry']['content']['event']['id'] = str(_uuid())
+    return request
+
+
+def add_event(event, tenant_id, region, log,
+              _prep_req=prepare_request):
+    """
+    Add event to cloud feeds
+    # Review: Is taking _prep_req as argument really required?
```

Comment:
Since `prepare_request` was not doing IO, I was thinking of leaving it as such but it unfortunately has random generation side-effect. Also, the request prepared was really quite big to to check in test. So, I decided to pass it as arg for easier testing.